File: backend/app/models/command.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Text, Enum as SQLEnum, ForeignKey
from app.core.types import UUID, JSONB
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class CommandType(str, enum.Enum):
    LOCK = "lock"
    UNLOCK = "unlock"
    WARNING = "warning"
    WIPE = "wipe"
    SYNC = "sync"
    UPDATE_POLICY = "update_policy"
    SHOW_MESSAGE = "show_message"
    HIDE_APP = "hide_app"
    UNHIDE_APP = "unhide_app"
    DISABLE_APP = "disable_app"
    ENABLE_APP = "enable_app"
    GPS_TRACK = "gps_track"
    CAMERA_ON = "camera_on"
    CAMERA_OFF = "camera_off"
    UNINSTALL_APP = "uninstall_app"
    SET_FRP_ACCOUNT = "set_frp_account"
    UPDATE_APP = "update_app"
    START_SCREEN_MIRROR = "start_screen_mirror"
    STOP_SCREEN_MIRROR = "stop_screen_mirror"
    START_AUDIO_STREAM = "start_audio_stream"
    STOP_AUDIO_STREAM = "stop_audio_stream"
    START_FILE_MANAGER = "start_file_manager"
    STOP_FILE_MANAGER = "stop_file_manager"

class CommandStatus(str, enum.Enum):
    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    EXECUTED = "executed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class DeviceCommand(Base):
    __tablename__ = "device_commands"
    
    id = Column(UUID(), primary_key=True, default=uuid.uuid4)
    
    device_id = Column(UUID(), ForeignKey("devices.id"), nullable=False)
    command_type = Column(SQLEnum(CommandType), nullable=False, index=True)
    
    # Command Payload
    payload = Column(JSONB, default={})
    
    # Status Tracking
    status = Column(SQLEnum(CommandStatus), default=CommandStatus.PENDING, index=True)
    retry_count = Column(String(10), default="0")
    max_retries = Column(String(10), default="3")
    
    # FCM Message ID
    fcm_message_id = Column(String(255))
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    sent_at = Column(DateTime)
    delivered_at = Column(DateTime)
    executed_at = Column(DateTime)
    
    # Error Info
    error_message = Column(Text)
    
    # Who issued the command
    issued_by = Column(UUID(), ForeignKey("users.id"))
    reason = Column(Text)
    
    # Relations
    device = relationship("Device", back_populates="commands")
    
    def __repr__(self):
        return f"<DeviceCommand {self.command_type} - {self.device_id}>"


class AuditLog(Base):
    __tablename__ = "audit_logs"
    
    id = Column(UUID(), primary_key=True, default=uuid.uuid4)
    
    # Entity Information
    entity_type = Column(String(50), nullable=False, index=True)  # device, customer, contract, etc.
    entity_id = Column(UUID(), index=True)
    
    # Action
    action = Column(String(100), nullable=False, index=True)
    
    # Changes
    old_value = Column(JSONB)
    new_value = Column(JSONB)
    
    # Who did it
    performed_by = Column(UUID(), ForeignKey("users.id"))
    ip_address = Column(String(50))
    user_agent = Column(Text)
    
    # Timestamp
    created_at = Column(DateTime, default=datetime.utcnow, index=True)
    
    def __repr__(self):
        return f"<AuditLog {self.entity_type} - {self.action}>"

File: backend-node/src/models/command.js
import { randomUUID } from 'crypto';
import { DataTypes } from 'sequelize';
import { sequelize } from '../db.js';
import { enumField, CommandType, CommandStatus } from '../utils/enums.js';

/**
 * Build a JSONB-compatible attribute. The old Python backend's JSONB type stored
 * JSON as TEXT on SQLite and native JSON/JSONB on Postgres/MySQL. We mirror that:
 * always (de)serialize through getters/setters so a TEXT column round-trips, and
 * native JSON columns also work (object passes through).
 */
function jsonbField(field, defaultValue) {
  return {
    type: DataTypes.TEXT,
    field,
    defaultValue: defaultValue === undefined ? undefined : JSON.stringify(defaultValue),
    get() {
      const raw = this.getDataValue(field);
      if (raw === null || raw === undefined) return raw;
      if (typeof raw === 'object') return raw; // native JSON column
      try {
        return JSON.parse(raw);
      } catch {
        return raw;
      }
    },
    set(val) {
      if (val === null || val === undefined) {
        this.setDataValue(field, val);
      } else if (typeof val === 'string') {
        this.setDataValue(field, val);
      } else {
        this.setDataValue(field, JSON.stringify(val));
      }
    },
  };
}

export const DeviceCommand = sequelize.define(
  'DeviceCommand',
  {
    id: { type: DataTypes.STRING(36), primaryKey: true, defaultValue: randomUUID },
    device_id: { type: DataTypes.STRING(36), allowNull: false },
    command_type: enumField({ field: 'command_type', allowNull: false }),
    payload: jsonbField('payload', {}),
    status: enumField({ field: 'status', defaultValue: 'pending' }),
    retry_count: { type: DataTypes.STRING(10), defaultValue: '0' },
    max_retries: { type: DataTypes.STRING(10), defaultValue: '3' },
    fcm_message_id: { type: DataTypes.STRING(255) },
    created_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    sent_at: { type: DataTypes.DATE },
    delivered_at: { type: DataTypes.DATE },
    executed_at: { type: DataTypes.DATE },
    error_message: { type: DataTypes.TEXT },
    issued_by: { type: DataTypes.STRING(36) },
    reason: { type: DataTypes.TEXT },
  },
  { tableName: 'device_commands', timestamps: false },
);

export const AuditLog = sequelize.define(
  'AuditLog',
  {
    id: { type: DataTypes.STRING(36), primaryKey: true, defaultValue: randomUUID },
    entity_type: { type: DataTypes.STRING(50), allowNull: false },
    entity_id: { type: DataTypes.STRING(36) },
    action: { type: DataTypes.STRING(100), allowNull: false },
    old_value: jsonbField('old_value'),
    new_value: jsonbField('new_value'),
    performed_by: { type: DataTypes.STRING(36) },
    ip_address: { type: DataTypes.STRING(50) },
    user_agent: { type: DataTypes.TEXT },
    created_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
  },
  { tableName: 'audit_logs', timestamps: false },
);

export { CommandType, CommandStatus };
export default DeviceCommand;
